File: Thrustline.WPF/Thrustline/Services/SimConnect/FlightDetector.cs
namespace Thrustline.Services.SimConnect;

/// <summary>
/// Flight state machine: idle → airborne → landed → idle (with 5s debounce).
/// Detects takeoff and landing transitions from SimConnect data.
/// </summary>
public class FlightDetector
{
    public event Action<string>? OnTakeoff; // departureIcao
    public event Action<FlightRecord>? OnLanding;

    private readonly AirportService _airports;
    private FlightState _state = FlightState.Idle;
    private bool? _prevOnGround;
    private double _prevVerticalSpeed;

    // Takeoff snapshot
    private double _fuelAtDeparture;
    private long _departureTime;
    private double _departureLat;
    private double _departureLon;

    // Debounce
    private long _landingLockUntil;

    public FlightState State => _state;

    public FlightDetector(AirportService airports)
    {
        _airports = airports;
    }

    public void Update(SimData data)
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

        // Skip if in debounce window
        if (now < _landingLockUntil)
        {
            _prevOnGround = data.SimOnGround;
            _prevVerticalSpeed = data.VerticalSpeed;
            return;
        }

        if (_prevOnGround.HasValue)
        {
            // Takeoff: was on ground → now airborne
            if (_prevOnGround.Value && !data.SimOnGround)
            {
                _state = FlightState.Airborne;
                _fuelAtDeparture = data.FuelQuantity;
                _departureTime = now;
                _departureLat = data.Latitude;
                _departureLon = data.Longitude;

                var depApt = _airports.FindNearest(_departureLat, _departureLon);
                OnTakeoff?.Invoke(depApt?.Icao ?? string.Empty);
            }

            // Landing: was airborne → now on ground
            if (!_prevOnGround.Value && data.SimOnGround && _state == FlightState.Airborne)
            {
                _state = FlightState.Landed;
                _landingLockUntil = now + 5000;

                var durationMin = (int)Math.Round((now - _departureTime) / 60_000.0);
                var fuelUsedGal = Math.Max(0, _fuelAtDeparture - data.FuelQuantity);
                var distanceNm = AirportService.HaversineNm(_departureLat, _departureLon, data.Latitude, data.Longitude);

                // Use previous tick's VS — current tick may already read 0
                var landingVsFpm = _prevVerticalSpeed;

                var depApt = _airports.FindNearest(_departureLat, _departureLon);
                var arrApt = _airports.FindNearest(data.Latitude, data.Longitude);

                var record = new FlightRecord
                {
                    DepartureIcao = depApt?.Icao ?? "UNKN",
                    ArrivalIcao = arrApt?.Icao ?? "UNKN",
                    DepartureLat = _departureLat,
                    DepartureLon = _departureLon,
                    ArrivalLat = data.Latitude,
                    ArrivalLon = data.Longitude,
                    DurationMin = Math.Max(1, durationMin),
                    FuelUsedGal = Math.Round(fuelUsedGal * 100) / 100,
                    DistanceNm = Math.Round(distanceNm * 10) / 10,
                    LandingVsFpm = Math.Round(landingVsFpm),
                };

                OnLanding?.Invoke(record);

                // Reset to idle after debounce
                Task.Delay(5000).ContinueWith(_ => _state = FlightState.Idle);
            }
        }

        _prevOnGround = data.SimOnGround;
        _prevVerticalSpeed = data.VerticalSpeed;
    }
}

File: legacy/electron/simconnect/flightDetector.ts
import type { SimData, FlightRecord, FlightState } from './types'
import { findNearestAirport } from '../../src/data/airports'

interface FlightCallbacks {
  onTakeoff: (departureIcao: string) => void
  onLanding: (record: FlightRecord) => void
}

interface FlightDetector {
  update: (data: SimData) => void
  getState: () => FlightState
}

/** Haversine distance in nautical miles */
function haversineNm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 3440.065 // Earth radius in nm
  const toRad = (deg: number) => (deg * Math.PI) / 180
  const dLat = toRad(lat2 - lat1)
  const dLon = toRad(lon2 - lon1)
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2
  return 2 * R * Math.asin(Math.sqrt(a))
}

export function createFlightDetector(callbacks: FlightCallbacks): FlightDetector {
  let state: FlightState = 'idle'
  let prevOnGround: boolean | null = null
  let prevVerticalSpeed = 0

  // Takeoff snapshot
  let fuelAtDeparture = 0
  let departureTime = 0
  let departureLat = 0
  let departureLon = 0

  // Debounce: ignore transitions for 5s after landing
  let landingLockUntil = 0

  function update(data: SimData): void {
    const now = Date.now()

    // Skip if in debounce window
    if (now < landingLockUntil) {
      prevOnGround = data.simOnGround
      prevVerticalSpeed = data.verticalSpeed
      return
    }

    // Detect transitions
    if (prevOnGround !== null) {
      // Takeoff: was on ground → now airborne
      if (prevOnGround && !data.simOnGround) {
        state = 'airborne'
        fuelAtDeparture = data.fuelQuantity
        departureTime = now
        departureLat = data.latitude
        departureLon = data.longitude
        const depApt = findNearestAirport(departureLat, departureLon)
        callbacks.onTakeoff(depApt?.icao ?? '')
      }

      // Landing: was airborne → now on ground
      if (!prevOnGround && data.simOnGround && state === 'airborne') {
        state = 'landed'
        landingLockUntil = now + 5000

        const durationMin = Math.round((now - departureTime) / 60_000)
        const fuelUsedGal = Math.max(0, fuelAtDeparture - data.fuelQuantity)
        const distanceNm = haversineNm(departureLat, departureLon, data.latitude, data.longitude)

        // Use previous tick's VS — current tick may already read 0
        const landingVsFpm = prevVerticalSpeed

        const depApt = findNearestAirport(departureLat, departureLon)
        const arrApt = findNearestAirport(data.latitude, data.longitude)

        const record: FlightRecord = {
          departureIcao: depApt?.icao ?? 'UNKN',
          arrivalIcao:   arrApt?.icao ?? 'UNKN',
          departureLat,
          departureLon,
          arrivalLat: data.latitude,
          arrivalLon: data.longitude,
          durationMin: Math.max(1, durationMin),
          fuelUsedGal: Math.round(fuelUsedGal * 100) / 100,
          distanceNm: Math.round(distanceNm * 10) / 10,
          landingVsFpm: Math.round(landingVsFpm),
        }

        callbacks.onLanding(record)

        // Reset to idle after debounce
        setTimeout(() => {
          state = 'idle'
        }, 5000)
      }
    }

    prevOnGround = data.simOnGround
    prevVerticalSpeed = data.verticalSpeed
  }

  return {
    update,
    getState: () => state,
  }
}
